### PR TITLE
release_collection.py - Add CHANGELOG handling

### PR DESCRIPTION
**Pull request by nhosoi**
- If a version update is detected in a role, retrieve the changes
  from the role's CHANGELOG.md and add them to COLLECTION_CHANGELOG.md.
- Copy COLLECTION_CHANGELOG.md to the converted collection's docs dir.

This pr depends upon https://github.com/linux-system-roles/auto-maintenance/pull/166 and https://github.com/linux-system-roles/auto-maintenance/pull/167.

@richm, this patch has not be tested as release_collection.py requires the version changes of one or more roles. Could you advice how you debugged the script? Maybe, I should use ci-testing for the purpose? Thanks! 

Note: this version has not supported when the ref is commit hash. Do we want to support a changelog item looks like this?
```
[COMMIT_HASH] - DATE
- change 0
- change 1
```

### CODE REVIEW

File: release_collection.py [hunk 1]
```diff
@@ -168,6 +169,28 @@ def process_ignore_and_lint_files(args, coll_dir):
         yaml.safe_dump(ansible_lint, open(os.path.join(coll_dir, ".ansible-lint"), "w"))
 
 
+def append_changelog(args, rolename, cur_ref, new_ref):
```
Comment (richm) [inline release_collection.py:172]: This function isn't appending anything, is it?  It looks like it is doing `get_role_changelog` or something like that.

File: release_collection.py [hunk 2]
```diff
@@ -382,6 +415,32 @@ def update_collection(args, galaxy, coll_rel):
...
 
+    # If to-be-appended changelogs are found, update COLLECTION_CHANGELOG.md
+    # and copy it to the collection docs dir.
+    if coll_changelog:
+        clhandle, clname = tempfile.mkstemp(suffix=".cl", prefix="collection")
+        with os.fdopen(clhandle, "w") as tf:
+            # Header
+            tf.write("Changelog\n=========\n\n")
+            # New changelogs
+            tf.write("{} - {}\n\n".format(galaxy["version"], datetime.now().date()))
```
Comment (richm) [inline release_collection.py:426]: What format does this `date()` function use?  If it is locale dependent, we should use a given, standard format string e.g. `YYYY-MM-DD`

**Comment by richm** [comment]
> * If a version update is detected in a role, retrieve the changes
>       from the role's CHANGELOG.md and add them to COLLECTION_CHANGELOG.md.
> 
>     * Copy COLLECTION_CHANGELOG.md to the converted collection's docs dir.
> 
> 
> This pr depends upon #166 and #167.
> 
> @richm, this patch has not be tested as release_collection.py requires the version changes of one or more roles. Could you advice how you debugged the script? Maybe, I should use ci-testing for the purpose? Thanks!

Can you edit collection_release.yml to put in whatever old versions you want?

> 
> Note: this version has not supported when the ref is commit hash. Do we want to support a changelog item looks like this?
> 
> ```
> [COMMIT_HASH] - DATE
> - change 0
> - change 1
> ```

I think let's get it working with tags first, then worry about getting it working with commit hashes.

File: release_collection.py (hunk 1 above)
Comment (nhosoi) [reply release_collection.py:172]: That's true. :p Sorry, I changed the original plan and forgot to adjust the function name...
Thanks!

File: release_collection.py (hunk 2 above)
Comment (nhosoi) [reply release_collection.py:426]: I thought it's the default behavior...
```
>>> from datetime import datetime
>>> print(datetime.now().date())
2022-07-14
```
It does not get affected by the env var LANG. Anything else we should be aware of?

File: release_collection.py [hunk 3]
```diff
@@ -80,6 +80,33 @@ def check_versions_updated(cur_ref, new_ref, versions_updated):
...
+def comp_versions(cur_ref, new_ref):
+    """Compare versions.
+
+    Return values:
+    1  - if cur_ref > new_ref
+    0  - if cur_ref == new_ref
+    -1 - if cur_ref < new_ref"""
+
+    try:
+        _mobj = re.match(r'v*(\d*.\d*.\d*)', cur_ref)
```
[the remark this answers is not in the document]
Comment (richm) [reply release_collection.py:92]: Does this mean the current function ``` has an error in that it doesn't handle versions from sshd like `v0.15.2`?
Note that the `StrictVersion` class has `_cmp` - so you can do
```
...
        if _mobj and _mobj.group(1):
            new_ref = _mobj.group(1)
        return StrictVersion(cur_ref)._cmp(StrictVersion(new_ref))
```
The reason why `check_versions_updated` iterates on the `version` array is because it has to know which version changed - the major, minor, or patch version, in order to do the semantic versioning for the collection. If you are just comparing two versions you can just use `_cmp`

File: release_collection.py [hunk 4]
```diff
@@ -340,6 +406,15 @@ def update_collection(args, galaxy, coll_rel):
                 check_versions_updated(
                     cur_ref, coll_rel[rolename]["ref"], versions_updated
                 )
+            # If a version update is detected, retrieve the new changelogs from CHANGELOG.md
+            if any(versions_updated):
```
Comment (richm) [inline release_collection.py:410]: You can't use `versions_updated` for individual roles - `versions_updated` keeps track of the semantic version changes to **all** roles (listed in `args.include`) - unfortunately `check_versions_updated` doesn't tell you if the version for the given role has changed - you'll have to modify it to do that (perhaps figure out a way to combine that with `comp_versions`

File: release_collection.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (nhosoi) [reply release_collection.py]: Thank you, @richm! I'm going to fix my code. BTW, when I try to play with StrictVersion, then I got this warning.
```
$ python --version
Python 3.10.5
$ python
...
>>> from distutils.version import StrictVersion
<stdin>:1: DeprecationWarning: The distutils package is deprecated and slated for removal in Python 3.12. Use setuptools or check PEP 632 for potential alternatives
```
It seems the new package `setuptools` does the same thing this way. Is it ok to replace `distutils` with `setuptools`?
```
from pkg_resources import parse_version
>>> parse_version("v1.2.3") > parse_version("1.2.2")
True
>>> parse_version("1.2.3") < parse_version("v1.2.2")
False
>>> parse_version("v1.2.3") == parse_version("1.2.2")
False
```

File: release_collection.py [hunk 5]
```diff
@@ -174,19 +201,34 @@ def get_role_changelog(args, rolename, cur_ref, new_ref):
...
             if cl.startswith("[{}]".format(new_ref)):
                 _print = True
-                _changelog = "{}\n{}".format(_changelog, cl)
+                _changelog = "{}\n#### {}".format(_changelog, cl)
             elif cl.startswith("[{}]".format(cur_ref)):
                 break
             elif _print:
-                _changelog = "{}\n{}".format(_changelog, cl)
+                if cl.lower() == "### new features":
+                    _changelog = "{}\n##### New Features".format(_changelog)
```
[the remark this answers is not in the document]
Comment (nhosoi) [reply release_collection.py:223]: This is further taken care in `compact_coll_changelog`. Yes, we actually do not care the count of # (changing H3 to H5) as long as it matches the condition in `compact_coll_changelog`. I just wanted the intermediate markdown be nice and clean. :)

File: release_collection.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (richm) [reply release_collection.py]: > It seems the new package setuptools does the same thing this way. Is it ok to replace distutils with setuptools?

tl;dr - let's do this in a separate PR - note that there is some other code in certificate and firewall that we will have to change too - let's just do it all at once - because we will have to write some custom code for this . . . :-( it seems `packaging` is not built-in to python 3.6
```
>python3.6
Python 3.6.15 (default, Jun 10 2022, 00:00:00) 
[GCC 11.3.1 20220421 (Red Hat 11.3.1-2)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from packaging.version import Version, parse
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ModuleNotFoundError: No module named 'packaging'
```
so I think we will have to create some sort of wrapper class or functions of our own for versioning e.g.
```python
try:
    import packaging.version

    def lsr_parse_version(v_str):
        return packaging.version.parse_version(v_str)

except ImportError:
    import distutils.version

    def lsr_parse_version(v_str):
        # strip leading "v" if any
        return distutils.version.StrictVersion(v_stripped)

def lsr_compare_versions(v1_str, v2_str):
    v1 = lsr_parse_version(v1_str)
    v2 = lsr_parse_version(v2_str)
    # no compare method :-(
    if v1 < v2: return -1
    elif v1 > v2: return 1
    else: return 0
```
Then `check_versions_updated` will be tricky because the array is `release` instead of `version` . . .

[the remark this answers is not in the document]
Comment (nhosoi) [reply release_collection.py]: Ok. I will leave the version handling in a separate PR (including the sshd's special case starting with `v`. Hope it's ok...). Thanks!

**Review by richm — DISMISSED** [review]
lgtm

**Comment by richm** [comment]
looks like you need to rebase

**Comment by nhosoi** [comment]
> looks like you need to rebase

Thank you, @richm. Done!

**Review by richm — APPROVED** [review]
